quickannotator/api/v1/utils/shared_crud.py

```diff
@@ -65,40 +45,181 @@ def get_annotation_query(model, scale_factor: float=1.0) -> Query:
+                "datetime": datetime.now()
+            })
+        stmt = insert(self.model).returning(self.model.id).values(new_annotations)
+        ids = db_session.scalars(stmt).all()
+        result = self.get_annotations_by_ids(annotation_ids=ids)
+        tile_ids = [ann.tile_id for ann in result]
+        
+        if self.is_gt:
+            upsert_tiles(self.image_id, self.annotation_class_id, tile_ids)
+        else:
+            upsert_tiles(self.image_id, self.annotation_class_id, tile_ids, pred_status=TileStatus.DONEPROCESSING)
+        return result
+
+
+    # READ
+    def get_annotation_by_id(self, annotation_id: int) -> models.Annotation:
+        result = get_annotation_query(self.model, 1/self.scaling_factor).filter_by(id=annotation_id).first()
+        return result
     
-    scale_factor = base_to_work_scaling_factor(image_id, annotation_class_id)
-    return get_annotation_query(model, scale_factor)
+    def get_annotations_by_ids(self, annotation_ids: List[int]) -> List[models.Annotation]:
+        result = get_annotation_query(self.model, 1/self.scaling_factor).filter(self.model.id.in_(annotation_ids)).all()
+        return result
+
+    def get_annotations_for_tiles(self, tile_ids: List[int]) -> List[models.Annotation]:
+        result = get_annotation_query(self.model, 1/self.scaling_factor).filter(self.model.tile_id.in_(tile_ids)).all()
+        return result
+
+    def get_annotations_within_poly(self, polygon: BaseGeometry) -> List[models.Annotation]:
+        scaled_polygon = self.scale_polygon(polygon, self.scaling_factor)
+        # NOTE: Sqlite may not use the spatial index here.
+        result = get_annotation_query(self.model, 1/self.scaling_factor).filter(func.ST_Intersects(self.model.polygon, func.ST_GeomFromText(scaled_polygon.wkt, 0))).all()
+        return result
+
+    # UPDATE
+    def update_annotation(self, annotation_id: int, polygon: BaseGeometry) -> models.Annotation:
+        scaled_polygon = self.scale_polygon(polygon, self.scaling_factor)
+        tile_id = get_tilespace(self.image_id, self.annotation_class_id).point_to_tileid(
+            scaled_polygon.centroid.x, scaled_polygon.centroid.y
+        )
+
+        annotation = db_session.query(self.model).filter(self.model.id == annotation_id).first()
+        
+        if annotation:
+            annotation.tile_id = tile_id
+            annotation.centroid = scaled_polygon.centroid.wkt
+            annotation.polygon = scaled_polygon.wkt
+            annotation.area = scaled_polygon.area
+            annotation.custom_metrics = compute_custom_metrics()
+            annotation.datetime = datetime.now()
+            db_session.commit()
+            return self.get_annotation_by_id(annotation_id)  # Get the updated annotation by id
+
+        return None  # Handle case where annotation_id is not found
+
+
+    # DELETE
+    def delete_annotation(self, annotation_id: int):
```

Comment:
i think there is a .delete function that supports the filter? : https://docs.sqlalchemy.org/en/20/tutorial/data_update.html#the-delete-sql-expression-construct